Allow setting custom volume sizes for Windows containers with containerd

marosset opened the pull request:
Signed-off-by: Mark Rossetti <marosset@microsoft.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind cleanup

<!--
Add one of the following kinds:
/kind bug
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
On Windows containers are created with a default volume size of 20Gb.
With docker this can be specified per-container with `--storage-opt "size=xxGb"` or globally by specifying `storage-opts` in the dockerd config file.

https://docs.microsoft.com/virtualization/windowscontainers/manage-containers/container-storage#storage-limits

This PR allows for the container volume size to be specified in pod specs and will pass that size to CRI container create calls.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Related to https://github.com/kubernetes/kubernetes/pull/108894 and https://github.com/containerd/containerd/issues/6694

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Allow setting custom root volume size for Windows containers with containerd v1.7+ when []containers.resources.limits.ephemeral-storage is specified.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

/sig windows
/area kubelet

marosset committed:
Allow setting custom volume sizes for Windows containers with contianerd

Signed-off-by: Mark Rossetti <marosset@microsoft.com>

marosset commented:
/hold
I will add some tests but will need the changes to pick up this new value implemented in containerd first

marosset commented:
/cc @dcantah

dcantah commented on pkg/kubelet/kuberuntime/kuberuntime_container_windows.go:
```diff
@@ -109,6 +109,17 @@ func (m *kubeGenericRuntimeManager) generateWindowsContainerConfig(container *v1
 		wc.Resources.MemoryLimitInBytes = memoryLimit
 	}
 
+	// Windows containers use ephermerral storage by default are are created with
+	// a default volume size of 20Gb.
+	// If specified - resources.limits.ephemeral-storage value will be used to to
+	// override this default.
```
```suggestion
	// Windows containers use ephemeral storage by default and are created with
	// a default volume size of 20GB.
	// If specified - resources.limits.ephemeral-storage value will be used to
	// override this default.
```

dcantah commented:
/lgtm

jsturtevant commented:
/triage accepted
/priority important-soon
/milestone v1.25

encodeflush commented:
Hi @marosset 
Bug triage team here!
It looks like this PR has a pending review. Just checking in to see if this is still on track for k8s 1.25.

marosset commented:
> Hi @marosset Bug triage team here! It looks like this PR has a pending review. Just checking in to see if this is still on track for k8s 1.25.

It is still on track for 1.25.
Thanks!

marosset committed:
Adding e2e test to validate setting rootfs size for Winows containers

marosset commented:
/test ?

marosset commented:
I added an e2e test to validate this behavior but the e2e test will skip unless the containerd version is v1.7 or higher.
I have done manual validation using https://github.com/kubernetes-sigs/sig-windows-tools/releases/download/windows-containerd-nightly/windows-containerd.tar.gz which builds containerd from main every night (and reports v1.7.0+unknown)

marosset commented:
/assign @jsturtevant @claudiubelu @dcantah

marosset commented:
/hold cancel

jsturtevant commented on pkg/kubelet/kuberuntime/kuberuntime_container_windows.go:
```diff
@@ -109,6 +109,17 @@ func (m *kubeGenericRuntimeManager) generateWindowsContainerConfig(container *v1
+	// override this default.
+	// https://docs.microsoft.com/virtualization/windowscontainers/manage-containers/container-storage#scratch-space
+	// https://docs.microsoft.com/virtualization/windowscontainers/manage-containers/container-storage#storage-limits
+	ephemeralStorageLimit := container.Resources.Limits.StorageEphemeral().Value()
```
Reading through https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/

```
If the kubelet is managing local ephemeral storage as a resource, then the kubelet measures storage use in:

emptyDir volumes, except tmpfs emptyDir volumes
directories holding node-level logs
writeable container layers

If a Pod is using more ephemeral storage than you allow it to, the kubelet sets an eviction signal that triggers Pod eviction.
```

It seems that kubelet tracks more than just the rootfs with the Ephemeral Storage Limit value.  The RootfssizeinBytes seems to be the "writable layer" here, but would this mean pods would be evicted if they added `emtpydir` volume as well as resize the rootsfs  of the container?

marosset replied:
@msau42 @ddebroy @mauriciopoppe - Do any of you (or anyone else from SIG-storage) have any thoughts here?

jsturtevant replied:
I think I am struggling a little bit with overloading the `ephemeralstoragelimit` concept and the `RootfsSizeInBytes`. I am not sure they are 1 to 1 but this is the first time I am looking at it.

I was looking around and it doesn't look like Linux sets the scratch layer sizes anywhere.  Does linux have a similar concept or is this windows only?

mauriciopoppe replied:
cc @jingxu97

dcantah replied:
The vhd itself for the volume is thin provisioned and will expand on data written, it's nowhere near 20GiB on the hosts disk

marosset replied:
> The vhd itself for the volume is thin provisioned and will expand on data written, it's nowhere near 20GiB on the hosts disk

That was what I thought - thanks for confirming!

dcantah commented on test/e2e/windows/volumes.go:
```diff
@@ -81,9 +87,11 @@ var _ = SIGDescribe("[Feature:Windows] Windows volume mounts ", func() {
-
 	})
 
+	ginkgo.It("validate rootfs size can be set larger than 20Gb", func() {
```
Curious, is there anybody that has a use case for it to be smaller that we know?

jsturtevant replied:
Is that a limitation in containerd only?  https://docs.microsoft.com/en-us/virtualization/windowscontainers/manage-containers/container-storage#storage-limits seems to suggest that it is possible in docker.

jsturtevant commented on test/e2e/windows/volumes.go:
```diff
@@ -188,3 +196,94 @@ func testPodWithROVolume(podName string, source v1.VolumeSource, path string) *v
+					Command: []string{
+						"powershell.exe",
+						"-Command",
+						"if (-not ((Get-PsDrive -Name C).Free -gt 21474836480)) { exit 1 } else { exit 0 }",
```
Ok, works for me, especially if the other layers are doing the validation

marosset commented:
@jingxu97 @ddebroy @mauriciopoppe - Can someone from SIG-storage (preferably with some Windows knowledge :P) weigh in here?

On Windows the default size for the 'writable container layer' is 20Gb and this cannot be expanded after the VM is created.
We have added a CRI field to be able to specify the size of the 'writable container layer' and updated containerd to consume this value and are looking for how best to expose this to users.
Does anyone have opinions on if we should map this value to the ephemeral storage requests of limits?
Since this value cannot be expanded after the container is created `limit` might make sense. The problem I see with using `limit` is that limits are used for determining when pods should be evicted which could be confusing to a user that wants to use a large portion of the writable container layer *and* and use things like `emptyDir` volumes which are also used when seeing if pods should get evicted.

/sig storage

jingxu97 commented:
> @jingxu97 @ddebroy @mauriciopoppe - Can someone from SIG-storage (preferably with some Windows knowledge :P) weigh in here?
> 
> On Windows the default size for the 'writable container layer' is 20Gb and this cannot be expanded after the VM is created. We have added a CRI field to be able to specify the size of the 'writable container layer' and updated containerd to consume this value and are looking for how best to expose this to users. Does anyone have opinions on if we should map this value to the ephemeral storage requests of limits? Since this value cannot be expanded after the container is created `limit` might make sense. The problem I see with using `limit` is that limits are used for determining when pods should be evicted which could be confusing to a user that wants to use a large portion of the writable container layer _and_ and use things like `emptyDir` volumes which are also used when seeing if pods should get evicted.
> 
> /sig storage

I also have concerns about overloading ephemeral storage request/limit for this usage. The request/limit is for the feature of local storage capacity isolation and have specific meaning about it. It already has some complexity on its own, I don't think we should overload more meaning on it. Right now this feature is beta and plan to promote to GA. But due to some system could not correctly get disk usage information (e.g., kind rootless), they have to disable this feature. 

For allow setting customizable writable container layer, do you think it is useful to set per pod/container level, or just cluster level?

marosset commented:
> > @jingxu97 @ddebroy @mauriciopoppe - Can someone from SIG-storage (preferably with some Windows knowledge :P) weigh in here?
> > On Windows the default size for the 'writable container layer' is 20Gb and this cannot be expanded after the VM is created. We have added a CRI field to be able to specify the size of the 'writable container layer' and updated containerd to consume this value and are looking for how best to expose this to users. Does anyone have opinions on if we should map this value to the ephemeral storage requests of limits? Since this value cannot be expanded after the container is created `limit` might make sense. The problem I see with using `limit` is that limits are used for determining when pods should be evicted which could be confusing to a user that wants to use a large portion of the writable container layer _and_ and use things like `emptyDir` volumes which are also used when seeing if pods should get evicted.
> > /sig storage
> 
> I also have concerns about overloading ephemeral storage request/limit for this usage. The request/limit is for the feature of local storage capacity isolation and have specific meaning about it. It already has some complexity on its own, I don't think we should overload more meaning on it. Right now this feature is beta and plan to promote to GA. But due to some system could not correctly get disk usage information (e.g., kind rootless), they have to disable this feature.
> 
> For allow setting customizable writable container layer, do you think it is useful to set per pod/container level, or just cluster level?

The best solution would allow for this to be set per-container but I think being able to set this per-pod would be acceptable.
I don't think that setting this per-cluster or per-node would be a good solution because the users who would want to set this might not have access to be able to do so.

encodeflush commented:
Hi @marosset 
Bug triage team here!
It looks like this PR has the pending review.
Just checking in to see if this is still on track for K8s 1.25.

marosset commented:
/milestone clear

marosset commented:
> Hi @marosset Bug triage team here! It looks like this PR has the pending review. Just checking in to see if this is still on track for K8s 1.25.

We are still having some on-going discussions on the best way to expose this field to users.
I don't think we'll reach agreement before 1.25 code freeze so I've removed this PR from the milestone

marosset commented:
/close
This needs more design/consideration to properly support.
Until then - the guidance SIG-Windows has for this is to use `emptydir` volumes.
